server-common/src/main/java/org/apache/kafka/common/DirectoryId.java

```diff
@@ -16,55 +16,166 @@
  */
 package org.apache.kafka.common;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
 import java.util.Set;
 
-public class DirectoryId {
+public class DirectoryId extends Uuid {
```

Comment:
I don't see why we should extend Uuid rather than just having a utility class that handles Uuids in the way that we want. Inheritance is almost always a mistake (unless it's of an interface)

This seems to add a lot of boilerplate to / from code compared with just using Uuid directly.

Reply:
I don't disagree with the general subclassing rule of thumb, but I can't see why it's bad on this case.

On the contrary, having a specific separate type helps enforce that the IDs are created from the `DirectoryId` class, not from the `UUID` class. This is useful because the names and reserved values are different. e.g. without a separate type, there are no compiler errors/warnings if:

* `Uuid.ZERO_UUID` is used to directly replace `DirectoryId.UNASSINGNED` creating confusion with the nomenclature
* `Uuid.randomUuid()` is used to directly replace `DirectoryId.random()`, silently ignoring the possibility that one of the reserved values may be incorrectly used. 

I don't think this is a huge deal either way though, so I you still feel strongly about this I'm ready – albeit somewhat reluctantly I'll admit – to change this and follow your suggestion.

Alternatively we could also keep a separate type, but use composition instead of inheritance. That will require a bit more boilerplate, and AFAICT bring little extra value, but we'd still get the benefits of having a separate type.

Reply:
It's a Uuid in every single way other than having 100 reserved entries at the beginning. I don't think that is enough to motivate a new type. A new type would be inconsistent with how we do everything else (cluster IDs, incarnation IDs, topic IDs, etc.) We use Uuid for all of them and there has never been any confusion.

(Although cluster ID is kind of a miss because technically it's a string and doesn't have to be a Uuid! But it was created before my time :) )

Having to copy between types is messy and confusing. For example, we'd have to copy between types every time we serialize to a record or to an RPC. This also removes a lot of the advantages you are touting (like "can't use Uuid.randomUuid" since we have to deal with Uuids anyway.)

Reply:
> we'd have to copy between types every time we serialize to a record or to an RPC. This also removes a lot of the advantages you are touting

No, I don't think so, because direct access to the metadata record fields is isolated to `PartitionRegistration` and `PartitionChangeBuilder`, so we'd get the benefits everywhere above that abstraction layer.

But I do agree with your point on inconsistency, so I've pushed [04a79cd](https://github.com/apache/kafka/pull/14290/commits/04a79cd4cec478e4cfece50c50fe4abd7f8c4e50).